### DISCUSSION TITLE
I2C question

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#11908] [Hardware & Peripherals]
I'm trying to work how to read data from an I2C submersible depth probe. After lots of messages about I2C being deprecated I ended up trying SoftI2C
```
import time; from machine import Pin,SoftI2C
i2c=SoftI2C(scl=Pin(10), sda=Pin(8), freq=1000)

print('i2c device addresses', i2c.scan())                                                                                                             # 40 for i2c depth probe
data=i2c.recv(3); print(data)
```
which gives
```
i2c device addresses [40]
Traceback (most recent call last):
  File "<stdin>", line 6, in <module>
AttributeError: 'SoftI2C' object has no attribute 'recv'
```
So the device is there at address 40 but I've got the receive part wrong. Can someone suggest the correct way to receive data please?

**Comment by davefes** [comment]
```py
>>> import umachine
>>> dir(umachine.SoftI2C)
['__class__', '__name__', 'readinto', 'start', 'stop', 'write', '__bases__', '__dict__', 'init', 'readfrom', 'readfrom_into', 'readfrom_mem', 'readfrom_mem_into', 'scan', 'writeto', 'writeto_mem', 'writevto']
```

**Comment by robert-hh** [comment]
What Dave want to tell. If you run print(dir(SoftI2C)) you get a list of methods that can be used. There is not recv. And with I2C you always have to tell the address of the device you want to communicate with. So for example:

SoftI2C.readfrom(40, 3)

See also the documentation: https://docs.micropython.org/en/latest/library/machine.I2C.html?highlight=softi2c#machine.SoftI2C

**Reply by kjm1102** [reply]
Tnx Rob. This probe came with no instructions other than the labels for the wires (3v3, gnd, sda, scl). If I expand the number of bytes to read from 3 to 5 to 9 to 19 etc it just sends me more more data b'\x06\x8aN', b'\x06\x8aN@\xd1', b'\x06\x8aN@\xd1I\xc8\x1a)', b"\x06\x8aN`\xd1I\xcc\x1a)9\x83E'0h\xa4\xe6\r\x14" when I'm just expecting a single depth reading of zero since the probe is not submerged.

Looks like the 5 bytes b'\x06\x8aN@\xd1' might be the sweet spot in that it's consistent. Does your internal data guru see any hint of a zero depth reading in there?  Big ask I know but I'm kinda out of my depth here (no pun intended).

**Reply by robert-hh** [reply]
DO you have a link or more information about the probe?

**Comment by kjm1102** [comment]
Purchased it from the inscrutable Chinese Rob, the only manufacturer I could find doing i2c depth probes (all the others are 4-20mA or RS485). Unfortunately no info on how to read it. I've lodged a ticket for some info on how to interrogate it but I'm not holding my breath. I was thinking if I could figure out what chip they've used I could dig up a data sheet on how to read it over i2C.

I'm fascinated that I can push the i2c frequency up to 30kHz & still get bytes out of it over a 5m length of cable.

**Reply by robert-hh** [reply]
With a 5m cable you'd better use RS485 instead of I2C.  Do you have a link to the product page at the `inscrutable Chinese`?

**Comment by kjm1102** [comment]
They sent me a couple of PDFs Rob, I think the one that doesn't mention temperatures or hibernation is maybe the correct one
[I2C Digital Output Pressure Sensor + Transmitter_00.pdf](https://github.com/micropython/micropython/files/11943242/I2C.Digital.Output.Pressure.Sensor.%2B.Transmitter_00.pdf)
[I2C Data Reading Instructions .pdf](https://github.com/micropython/micropython/files/11943228/I2C.Data.Reading.Instructions.pdf)

**Comment by robert-hh** [comment]
That sheds a little bit of light on the protocol, even if it is inconsistent. The sequence depends in the hibernating state of the sensor.
- If it is operational, you just read 4 bytes, with the first tow bytes being the pressure and the next are the temperature.
- If it is hibernating, you first have to write 0 bytes of data to the device to wake it up, then wait 8-10 ms, then read 4 bytes of data as above.

It may be that the second method cal always be used.

**Reply by kjm1102** [reply]
Tnx Rob, I tried the operational one because it looked easier
```
import time; from machine import Pin,SoftI2C
i2c=SoftI2C(scl=Pin(10), sda=Pin(8), freq=10000)                                                                                                      # works up to freq=33k

addr=i2c.scan(); print('i2c device addresses', addr)                                                                                                  # 40h for i2c depth probe
data=i2c.readfrom(40, 4) if addr else 'no probe'                                                                                                      # read 4 bytes
Pb=data[:2]; Tb=data[2:]; print('data',data, ' Pbytes',Pb, ' Tbytes', Tb)                                                                             # 1st are pressure, next 2 are temperature
Pv=int(Pb.hex(),16); Tv=int(Tb.hex(),16); print('Pv',Pv, ' Tv',Tv)                                                                                    # convert to hex then deciml
print('temperature(C)', 200*Tv/2047-50)

>>> 
i2c device addresses [40]
data b'\x06\x8aW`'  Pbytes b'\x06\x8a'  Tbytes b'W`'
Pv 1674  Tv 22368
temperature(C) 2135.442
```
but I seem to be out by 2 orders of magnitude. Can you see where I'm going wrong?

**Comment by robert-hh** [comment]
According to the data pattern in figure 2, you have to shift the Tv value by 5 bits to the right, which is  divide by 32, before conversion. Then you get 18.29 °C.

**Reply by kjm1102** [reply]
When you say 'before conversion' do you mean before using the 200*Tv/2047-50 formula?
```
import time; from machine import Pin,SoftI2C
i2c=SoftI2C(scl=Pin(10), sda=Pin(8), freq=10000)                                                                                                      # works up to freq=33k

addr=i2c.scan(); print('i2c device addresses', addr)                                                                                                  # 40h for i2c depth probe
#buf=b''; i2c.writeto(40, buf, stop=True); time.sleep(.009))
#buf=b''; i2c.write(buf); time.sleep(.009)
data=i2c.readfrom(40, 4) if addr else 'no probe'                                                                                                      # read 4 bytes
Pb=data[:2]; Tb=data[2:]; print('data',data, ' Pbytes',Pb, ' Tbytes', Tb)                                                                             # 1st 2 bytes are pressure, next 2 are temperature
Pv=int(Pb.hex(),16); Tv=int(Tb.hex(),16)/32; print('Pv',Pv, ' Tv',Tv)                                                                                 # convert to hex then deciml
print('temperature(C)', 200*Tv/2047-50)

i2c device addresses [40]
data b'\x06\x8aW '  Pbytes b'\x06\x8a'  Tbytes b'W '
Pv 1674  Tv 697.0
temperature(C) 18.09966
```
In which case I get 18.10 not 18.29

**Comment by robert-hh** [comment]
> When you say 'before conversion' do you mean before using the 200*Tv/2047-50 formula?

Yes. You could use integer divide as well or shift by 5 bit to the right. Like `Tv=int(Tb.hex(),16) // 32` or `Tv=int(Tb.hex(),16) >> 5`

> In which case I get 18.10 not 18.29

I got 18.29 when starting with the raw value of 22368 of your first post. The values vary always a little bit. +/- 1 is the systematic error when taking digital values, and then there is the noise of the sensor. The question I have is, whether ~18°C is reasonable for the location of the sensor.

**Reply by kjm1102** [reply]
Yeah 18 degrees C seems reasonable for the temperature here on the test bench. I don't think I'm out of the woods though, the more I look at that .pdf the more I think I need to be filtering at the bit level rather than bytes. With pressure for example I feel I should be checking bits 15 & 14, I'm thinking those 2 status bits might give me a clue as to what mode the sensor is in. I haven't been able to figure out how to get it into hibernate mode.
